Library/OcVirtualFsLib/VirtualDir.c

```diff
@@ -0,0 +1,586 @@
+/** @file
+  Copyright (C) 2020, Goldfish64. All rights reserved.
+
+  All rights reserved.
+
+  This program and the accompanying materials
+  are licensed and made available under the terms and conditions of the BSD License
+  which accompanies this distribution.  The full text of the license may be found at
+  http://opensource.org/licenses/bsd-license.php
+
+  THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+  WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+**/
+
+#include <Uefi.h>
+
+#include <Library/BaseLib.h>
+#include <Library/BaseMemoryLib.h>
+#include <Library/DebugLib.h>
+#include <Library/MemoryAllocationLib.h>
+#include <Library/UefiBootServicesTableLib.h>
+#include <Library/UefiRuntimeServicesTableLib.h>
+#include <Library/OcGuardLib.h>
+#include <Library/OcFileLib.h>
+#include <Library/OcVirtualFsLib.h>
+
+#include <Guid/FileInfo.h>
+
+#include "VirtualFsInternal.h"
+
+STATIC
+EFI_STATUS
+EFIAPI
+VirtualDirOpen (
+  IN  EFI_FILE_PROTOCOL       *This,
+  OUT EFI_FILE_PROTOCOL       **NewHandle,
+  IN  CHAR16                  *FileName,
+  IN  UINT64                  OpenMode,
+  IN  UINT64                  Attributes
+  )
+{
+  EFI_STATUS         Status;
+  VIRTUAL_DIR_DATA   *Data;
+
+  ASSERT (This != NULL);
```

Comment:
Please do not use `ASSERT` for protocol functions, as we want the checks to be valid in RELEASE builds and actually return an error — `EFI_INVALID_PARAMETER`. This is an EDK II convention at least.

This applies to all the functions below.